ksmt-cvc5/src/main/kotlin/org.ksmt.solver.cvc5/KCvc5ExprInternalizer.kt

```diff
@@ -1057,29 +1098,146 @@ class KCvc5ExprInternalizer(
+        isUniversal: Boolean
+    ): E = transform(body) { transformedBody: Term ->
+        val boundConsts = bounds.map { it.internalizeDecl() }
+        mkQuantifierTerm(isUniversal, boundConsts, transformedBody)
+    }
 
-            val cvc5Vars = cvc5Consts.map { nsolver.mkVar(it.sort, it.symbol) }.toTypedArray()
-            val cvc5BodySubstituted = cvc5Body.substitute(cvc5Consts, cvc5Vars)
+    private fun mkQuantifierTerm(isUniversal: Boolean, bounds: List<Term>, body: Term): Term =
+        resolveBoundVars(bounds, body) { boundVars, bodyWithVars ->
+            nsolver.mkQuantifier(isUniversal, boundVars, bodyWithVars)
+        }
 
-            nsolver.mkQuantifier(isUniversal, cvc5Vars, cvc5BodySubstituted)
+    private fun mkLambdaTerm(bounds: List<Term>, body: Term): Term =
+        resolveBoundVars(bounds, body) { boundVars, bodyWithVars ->
+            nsolver.mkLambda(boundVars, bodyWithVars)
         }
+
+    private inline fun resolveBoundVars(bounds: List<Term>, body: Term, mkTerm: (Array<Term>, Term) -> Term): Term {
+        val boundVars = bounds.map {
+            if (it.hasSymbol()) {
+                nsolver.mkVar(it.sort, it.symbol)
+            } else {
+                nsolver.mkVar(it.sort)
+            }
+        }.toTypedArray()
+
+        val bodyWithVars = body.substitute(bounds.toTypedArray(), boundVars)
+        return mkTerm(boundVars, bodyWithVars)
     }
 
-    override fun transform(expr: KExistentialQuantifier) = transformQuantifier(expr, isUniversal = false)
+    private fun Term.mkFunctionApp(args: List<Term>): Term =
+        nsolver.mkTerm(Kind.APPLY_UF, arrayOf(this) + args)
 
-    override fun transform(expr: KUniversalQuantifier) = transformQuantifier(expr, isUniversal = true)
+    private fun mkAndTerm(args: List<Term>): Term =
+        if (args.size == 1) args.single() else nsolver.mkTerm(Kind.AND, args.toTypedArray())
 
-    override fun <A : KArraySortBase<R>, R : KSort> transform(expr: KFunctionAsArray<A, R>): KExpr<A> {
-        throw KSolverUnsupportedFeatureException(
-            "No direct impl in cvc5 (as-array is CONST_ARRAY term with base array element)"
-        )
+    private fun mkOrTerm(args: List<Term>): Term =
+        if (args.size == 1) args.single() else nsolver.mkTerm(Kind.OR, args.toTypedArray())
+
+    private fun mkArraySelectTerm(array: Term, indices: List<Term>): Term =
+        if (array.sort.isArray) {
+            val selectIdx = mkArrayOperationIndex(indices)
+            nsolver.mkTerm(Kind.SELECT, array, selectIdx)
+        } else {
+            nsolver.mkTerm(Kind.APPLY_UF, arrayOf(array) + indices)
+        }
+
+    private fun mkArrayStoreTerm(array: Term, indices: List<Term>, value: Term): Term =
+        if (array.sort.isArray) {
+            val storeIdx = mkArrayOperationIndex(indices)
+            nsolver.mkTerm(Kind.STORE, arrayOf(array, storeIdx, value))
+        } else {
+            val boundVars = indices.map { nsolver.mkVar(it.sort) }
```

Comment:
Possible symbols for indices don't matter? Or they cannot have ones?

Reply:
Here we are creating fresh bound vars and therefore they don't have symbols. Vars can only have a symbol if they were `bound vars` in KSMT. This is only possible for `lambda expressions` and `quantifiers`.